src/uct/ib/base/ib_pd.c

```diff
@@ -0,0 +1,466 @@
+                        uct_mem_h *memh_p UCS_MEMTRACK_ARG)
+{
+    uct_ib_rcache_region_t *region;
+    ucs_status_t status;
+
+    region = ucs_calloc(1, sizeof(*region), "uct_ib_region");
+    if (region == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    status = uct_ib_mem_alloc(uct_pd, length_p, address_p,
+                              (uct_mem_h *)&region->mr UCS_MEMTRACK_VAL);
+    if (status != UCS_OK) {
+        ucs_free(region);
+        return status;
+    }
+
+    region->super.super.start = (uintptr_t)*address_p;
+    region->super.super.end   = (uintptr_t)*address_p + *length_p;
+
+    uct_ib_rcfache_region_set_stub_mr(region);
+    *memh_p = &region->stub_mr;
+    return UCS_OK;
+}
+
+static ucs_status_t uct_ib_mem_rcache_free(uct_pd_h uct_pd, uct_mem_h memh)
+{
+    uct_ib_rcache_region_t *region = uct_ib_rache_region_from_memh(memh);
+    ucs_status_t status;
+
+    status = uct_ib_mem_free(uct_pd, region->mr);
+    ucs_free(region);
+    return status;
+}
+
+static ucs_status_t uct_ib_mem_rcache_reg(uct_pd_h uct_pd, void *address,
+                                          size_t length, uct_mem_h *memh_p)
+{
+    uct_ib_pd_t *pd = ucs_derived_of(uct_pd, uct_ib_pd_t);
+    ucs_rcache_region_t *rregion;
+    ucs_status_t status;
+
+    status = ucs_rcache_get(pd->rcache, address, length, PROT_READ|PROT_WRITE,
+                            &rregion);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucs_assert(rregion->refcount > 0);
+    *memh_p = &ucs_derived_of(rregion, uct_ib_rcache_region_t)->stub_mr;
+    return UCS_OK;
+}
+
+static ucs_status_t uct_ib_mem_rcache_dereg(uct_pd_h uct_pd, uct_mem_h memh)
+{
+    uct_ib_pd_t *pd = ucs_derived_of(uct_pd, uct_ib_pd_t);
+    uct_ib_rcache_region_t *region = uct_ib_rache_region_from_memh(memh);
+
+    ucs_rcache_region_put(pd->rcache, &region->super);
+    return UCS_OK;
+}
+
+static uct_pd_ops_t uct_ib_pd_rcache_ops = {
+    .close        = uct_ib_pd_close,
+    .query        = uct_ib_pd_query,
+    .mem_alloc    = uct_ib_mem_rcache_alloc,
+    .mem_free     = uct_ib_mem_rcache_free,
+    .mem_reg      = uct_ib_mem_rcache_reg,
+    .mem_dereg    = uct_ib_mem_rcache_dereg,
+    .mkey_pack    = uct_ib_mkey_pack,
+};
+
+
+static ucs_status_t uct_ib_rcache_mem_reg_cb(void *context, ucs_rcache_t *rcache,
+                                             ucs_rcache_region_t *rregion)
+{
+    uct_ib_rcache_region_t *region = ucs_derived_of(rregion, uct_ib_rcache_region_t);
+    uct_ib_pd_t *pd = context;
+    ucs_status_t status;
+
+    status = uct_ib_mem_reg(&pd->super, (void*)region->super.super.start,
+                            region->super.super.end - region->super.super.start,
+                            (uct_mem_h *)&region->mr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    uct_ib_rcfache_region_set_stub_mr(region);
+    return UCS_OK;
+}
+
+static void uct_ib_rcache_mem_dereg_cb(void *context, ucs_rcache_t *rcache,
```

Comment:
The de-registration may result in an error. Would it make sense to propagate the error to the upper layer ?